interfaces/Readme.md

```diff
@@ -0,0 +1,482 @@
+# CSC207 Course Notes - Interfaces
+
+A key part of Java's class design are interfaces.
```

Comment:
Maybe something introducing the ways that interfaces are used? (This might be too long-winded, or maybe even off topic.)

There are (at least) 3 ways that Java interfaces are used.

1. A class provides a service that operates on data supplied by a client programmer.

For example, class [`java.util.Arrays`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Arrays.html) "contains various methods for manipulating arrays (such as sorting and searching)." In particular, it has [a `sort` method](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Arrays.html#sort(java.lang.Object%5B%5D)).

This `sort` method was written so that it can sort _anything_ that can be compared — even for classes haven't been created yet! This is possible because of interfaces: the Javadoc for method `sort` specifies that "All elements in the array must implement the [Comparable](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html) interface." That means that they have a `compareTo` method, which is called inside the sorting code.

If you write code that manipulates data, consider making an interface describing what behaviours you need the data to have.

2. When there are multiple ways to implement a concept.

For example, a list of things could be stored in an array, a wrapper class that contains an array (like [`java.util.ArrayList`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ArrayList.html)), or a wrapper class that contains a linked list of nodes (like [java.util.LinkedList](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/LinkedList.html)). Each of these choices has benefits and limitations.

3. To cross an architectural boundary (which often feels like a combination of the previous two points).

For example, in Clean Architecture, a _use case_ is independent from the user interface. Changing to a different user interface (from Java Swing to a web-based application, perhaps) shouldn't affect the use case code.

Here, the use case programmer will create an interface that describes how to call the use case: the parameters will contain the data needed by the use case — and it's the job of the user interface to gather that data before calling the use case.

Reply:
Maybe this belongs further down?

Reply:
I'd be happy to add this maybe at the end of the document as a 'when do we use interfaces' sort of section. I think it's too much to have before the examples as it benefits from understanding what the interface looks like in action. Definitely the points about 'what properties data have' is key to using interfaces efficiently.